include/drivers/pinctrl.h

```diff
@@ -0,0 +1,211 @@
+ * @PIN_CONFIG_END: this is the last enumerator for pin configurations, if
+ *	you need to pass in custom configurations to the pin controller, use
+ *	PIN_CONFIG_END+1 as the base offset.
+ * @PIN_CONFIG_MAX: this is the maximum configuration value that can be
+ *	presented using the packed format.
+ */
+enum pin_config_param {
+	PIN_CONFIG_BIAS_BUS_HOLD,
+	PIN_CONFIG_BIAS_DISABLE,
+	PIN_CONFIG_BIAS_HIGH_IMPEDANCE,
+	PIN_CONFIG_BIAS_PULL_DOWN,
+	PIN_CONFIG_BIAS_PULL_PIN_DEFAULT,
+	PIN_CONFIG_BIAS_PULL_UP,
+	PIN_CONFIG_DRIVE_OPEN_DRAIN,
+	PIN_CONFIG_DRIVE_OPEN_SOURCE,
+	PIN_CONFIG_DRIVE_PUSH_PULL,
+	PIN_CONFIG_DRIVE_STRENGTH,
+	PIN_CONFIG_DRIVE_STRENGTH_UA,
+	PIN_CONFIG_INPUT_DEBOUNCE,
+	PIN_CONFIG_INPUT_ENABLE,
+	PIN_CONFIG_INPUT_SCHMITT,
+	PIN_CONFIG_INPUT_SCHMITT_ENABLE,
+	PIN_CONFIG_MODE_LOW_POWER,
+	PIN_CONFIG_MODE_PWM,
+	PIN_CONFIG_OUTPUT,
+	PIN_CONFIG_OUTPUT_ENABLE,
+	PIN_CONFIG_OUTPUT_IMPEDANCE_OHMS,
+	PIN_CONFIG_PERSIST_STATE,
+	PIN_CONFIG_POWER_SOURCE,
+	PIN_CONFIG_SKEW_DELAY,
+	PIN_CONFIG_SLEEP_HARDWARE_STATE,
+	PIN_CONFIG_SLEW_RATE,
+	PIN_CONFIG_END = 0x7F,
+	PIN_CONFIG_MAX = 0xFF,
+};
+
+/*
+ * Helpful configuration macro to be used in tables etc.
+ */
+#define PIN_CONF_PACKED(cfg_id, arg)                                           \
+	(((arg) << 8) | (((uint32_t)cfg_id) & 0xffUL))
+
+/*
+ * The following inlines stuffs a configuration parameter and data value
+ * into and out of an unsigned long argument, as used by the generic pin config
+ * system. We put the parameter in the lower 8 bits and the argument in the
+ * upper 24 bits.
+ */
+
+static inline enum pin_config_param
+pinconf_to_config_param(unsigned int config)
+{
+	return (enum pin_config_param)(config & 0xffUL);
+}
+
+static inline uint32_t pinconf_to_config_argument(uint32_t config)
+{
+	return (uint32_t)((config >> 8) & 0xffffffUL);
+}
+
+static inline uint32_t
+pinconf_to_config_packed(enum pin_config_param param, uint32_t argument)
+{
+	return PIN_CONF_PACKED(param, argument);
+}
+
+#define PINCTRL_PIN(pin_number, pin_name)                                      \
+	{                                                                      \
+		.pin = pin_number, .name = pin_name                            \
+	}
+
+struct pinctrl_pin {
+	uint32_t pin;
+	uint32_t configs;
+	const char *name;
+};
+
+/* pinctrl functions */
+uint32_t pinctrl_get_groups_count(void);
+const char *pinctrl_get_group_name(uint32_t group_id);
+int pinctrl_get_group_pins(uint32_t group_id, const uint32_t **pins,
+			   uint32_t *num_pins);
+
+uint32_t pinctrl_get_pins_count(void);
+const char *pinctrl_get_pin_name(uint32_t pin_id);
+int pinctrl_get_pins(const struct pinctrl_pin **pins, uint32_t *num_pins);
+
+uint32_t pinctrl_get_functions_count(void);
+const char *pinctrl_get_function_name(uint32_t func_id);
+int pinctrl_get_function_groups(uint32_t func_id, const uint32_t **groups,
+				uint32_t *num_groups);
+
+/* pinmux functions */
+int pinctrl_request_group(uint32_t group_id);
```

Comment:
What are the purpose for the next 4 functions? As I can see, you do not implement them in BRCM2712 driver.

Reply:
It's called by Linux agent by using PINCTRL_REQUEST. The BRCM2712 has nothing to do, so it implements stubs which always succeed.
More over, for reason unknown to me, Linux agent accessing both pin_groups and pins, and expect both of them to be processed without failure. Therefor  BRCM2712 is implemented considered group_id  == pin_id (the same way as Linux does).